### ISSUE TITLE
make sure pylint tests are load balanced / not over-duplicated across travis jobs

### ISSUE THREAD

**Issue by goodfeli** [cleverhans-lab/cleverhans#602]

**Comment by goodfeli** [comment]
Is it any faster if we do one command line call and pass all the files in one call? Does this prevent repeated imports?

**Comment by goodfeli** [comment]
Is there a way to configure it to do shallower analysis where it doesn't need to import everything?

**Comment by goodfeli** [comment]
I'm not sure yet how we would keep track of this kind of persistent state on travis, but maybe we could use a caching system, where we only run pylint on a file if the md5sum of the file concatenated with pylintrc has changed since the last time it passed pylint.

**Comment by goodfeli** [comment]
Previous comment is wrong: we can't hash an individual file, because other files might have changed interfaces.

e.g. if file A says:
import file_b
foo()

but file_b now has
def foo(x):

file A is wrong not to supply an argument to foo.

**Comment by goodfeli** [comment]
Can we speed it up with multithreading? Everything runs in a subprocess anyway. I/O is probably a reasonable amount of the work so using multiple processes may help even on a single core.

**Comment by goodfeli** [comment]
It's slow locally too

**Comment by goodfeli** [comment]
I'm looking into this a bit already.

One thing I've learned: we want to call pylint on multiple files for correctness, not just speed. Run on multiple files at once, it can detect additional errors, like cyclic imports and code duplicated across files.

Also, tentatively, it looks like there is a speed benefit. On my laptop I ran pylint on two different files. It takes about 10s to run each file individually, but it still takes about 10s to run both files in one call. I think the runtime is dominated by importing tensorflow. I need to check that this scales to multiple files, travis, etc.

**Comment by npapernot** [comment]
Ok one thing is that it might make sense for us to move to something with a bit more ressources like https://github.com/marketplace/google-cloud-build if Travis starts to be limited?

**Comment by goodfeli** [comment]
I don't think that google-cloud-build would speed up pylint in particular because pylint turns out not to be very parallelizable. I'll open a separate issue to look into using Google Cloud Build

**Comment by goodfeli** [comment]
After #603 it might be possible to close this issue, but I want to check a few things. It looks like travis tests aren't load balanced very well. If we look into it more and that's accurate, the pylint tests might be a big part of that. We might want to divide them up to different travis workers. Other than that I don't think this is low hanging fruit anymore.

**Comment by jonasguan** [comment]
Closing as we upgrade to version 4 and switch to Github Actions.
